Std/Lean/Meta/UnusedNames.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2022 Jannis Limperg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jannis Limperg
+-/
+
+import Lean.Meta
+import Std.Data.String
+
+open Lean Lean.Meta
+
+namespace Lean.Name
+
+private def parseIndexSuffix (s : Substring) : Option Nat :=
+  if s.isEmpty then
+    none
+  else if s.front == '_' then
+    s.drop 1 |>.toNat?
+  else
+    none
+
+/--
+Result type of `Lean.Name.matchUpToIndexSuffix`. See there for details.
+-/
+inductive MatchUpToIndexSuffix
+  /-- Exact match. -/
+  |  exactMatch
+  /-- No match. -/
+  | noMatch
+  /-- Match up to suffix. -/
+  | suffixMatch (i : Nat)
+
+/--
+Succeeds if `n` is equal to `query`, except `n` may have an additional `_i`
+suffix for some natural number `i`. More specifically:
+
+- If `n = query`, the result is `exactMatch`.
+- If `n = query ++ "_i"` for some natural number `i`, the result is
+  `suffixMatch i`.
+- Otherwise the result is `noMatch`.
+-/
+def matchUpToIndexSuffix (n : Name) (query : Name) :
+    MatchUpToIndexSuffix :=
+  match n, query with
+  | .str pre₁ s₁, .str pre₂ s₂ =>
+    if pre₁ != pre₂ then
+      .noMatch
+    else
+      if let some suffix := s₁.dropPrefix? s₂ then
+        if suffix.isEmpty then
+          .exactMatch
+        else
+          if let some i := parseIndexSuffix suffix then
+            .suffixMatch i
+          else
+            .noMatch
+      else
+        .noMatch
+  | n, query => if n == query then .exactMatch else .noMatch
+
+end Name
+
+
+namespace LocalContext
+
+/--
+Obtain the least natural number `i` such that `suggestion ++ "_i"` is an unused
+name in the given local context. If `suggestion` itself is unused, the result
+is `none`.
+-/
+def getUnusedUserNameIndex (lctx : LocalContext) (suggestion : Name) :
+    Option Nat := Id.run do
+  let mut minSuffix := none
+  for ldecl in lctx do
+    let hypName := ldecl.userName
+    if hypName.hasMacroScopes then
+      continue
+    match ldecl.userName.matchUpToIndexSuffix suggestion with
+    | .exactMatch => minSuffix := updateMinSuffix minSuffix 1
+    | .noMatch => continue
+    | .suffixMatch i => minSuffix := updateMinSuffix minSuffix (i + 1)
+  minSuffix
+  where
```

Comment:
You might have missed it in the other code suggestion but these where blocks should be dedented so that `def` lines up with `where` (and  `termination_by` when applicable).